### DISCUSSION TITLE
Is it possible to pass a failure callback to a job when queueing it?

### DISCUSSION THREAD

**Discussion by jameshulse** [laravel/framework#47410] [Q&A] [answered]
I'm aware that you can implement the `failed()` function to run if the job fails, but is there a way to pass a failure callback from the caller e.g:

```
MyJob::dispatch()
    ->catch(fn () => /* ...do some clean up */);
```

I expect that my job may be called in different contexts, and may want to clean up based on that context, rather than having a generic cleanup.

An example is a job to send an email when something is created. The first time it runs, if it fails, I would want to change the status on the model back to "draft" or "failed". But, if I allow the user to resend the email at a later time and this fails, the model has already been created so I don't want to change the status on failure. I would much prefer that this logic was baked into the actions that allow for sending, and not the job itself knowing the context that it is called in.

I hope that makes sense!

**Comment by henzeb** [comment, marked as answer]
It's possible with batches. Not with a single Job. But you can incorporate the logic yourself in the Job. Just pass your closure in to the dispatch method, which will pass it to your constructor.
